Broken optionals in Python 3.9

### Describe the bug
Following #2030 (86f761d7919fcd8f536882e7a2f1538d1793179a) the following test is failing for Python 3.9 only:
```
_________________________________ test_f11[c] __________________________________
[gw0] linux -- Python 3.9.20 /opt/hostedtoolcache/Python/3.9.20/x64/bin/python

Module_5 = (<module 'modules.Module_5' from '/home/runner/work/pyccel/pyccel/tests/epyccel/modules/Module_5.py'>, <module 'Module.../runner/work/pyccel/pyccel/tests/epyccel/modules/__epyccel__gw0/Module_5_jujr745n8d1s.cpython-39-x86_64-linux-gnu.so'>)

    def test_f11(Module_5):
        mod, modnew = Module_5
    
        # ...
        assert mod.call_optional_11() == modnew.call_optional_11()
>       assert mod.call_optional_12() == modnew.call_optional_12()
E       AssertionError: assert 9 == 0
E        +  where 9 = <function call_optional_12 at 0x7ff811f785e0>()
E        +    where <function call_optional_12 at 0x7ff811f785e0> = <module 'modules.Module_5' from '/home/runner/work/pyccel/pyccel/tests/epyccel/modules/Module_5.py'>.call_optional_12
E        +  and   0 = <built-in function call_optional_12>()
E        +    where <built-in function call_optional_12> = <module 'Module_5_jujr745n8d1s' from '/home/runner/work/pyccel/pyccel/tests/epyccel/modules/__epyccel__gw0/Module_5_jujr745n8d1s.cpython-39-x86_64-linux-gnu.so'>.call_optional_12

epyccel/test_epyccel_optional_args.py:147: AssertionError
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -75,6 +75,7 @@ All notable changes to this project will be documented in this file.
 -   #1972 : Simplified `printf` statement for Literal String.
 -   #2026 : Fix missing loop in slice assignment.
 -   #2008 : Ensure list/set/dict assignment is recognised as a reference.
+-   #2039 : Ensure expressions in the iterable of a for loops is calculated before the loop.
 
 ### Changed
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -75,7 +75,7 @@ All notable changes to this project will be documented in this file.
 -   #1972 : Simplified `printf` statement for Literal String.
 -   #2026 : Fix missing loop in slice assignment.
 -   #2008 : Ensure list/set/dict assignment is recognised as a reference.
--   #2039 : Ensure expressions in the iterable of a for loops is calculated before the loop.
+-   #2039 : Ensure any expressions in the iterable of a for loop are calculated before the loop.
 
 ### Changed
 

--- a/pyccel/codegen/printing/ccode.py
+++ b/pyccel/codegen/printing/ccode.py
@@ -2362,6 +2362,10 @@ class CCodePrinter(CodePrinter):
                 stop_condition = f'({step_code} > 0) ? ({counter} < {stop_code}) : ({counter} > {stop_code})'
             for_code = f'for ({counter} = {start_code}; {stop_condition}; {counter} += {step_code})\n'
 
+        if self._additional_code:
+            for_code = self._additional_code + for_code
+            self._additional_code = ''
+
         body = self._print(additional_assign) + self._print(expr.body)
 
         self.exit_scope()

